core/chains/solana/chain.go

```diff
@@ -0,0 +1,189 @@
+	}
+	tc := func() (solanaclient.ReaderWriter, error) {
+		return ch.getClient()
+	}
+	ch.txm = soltxm.NewTxm(tc, cfg, lggr)
+	ch.balanceMonitor = monitor.NewBalanceMonitor(ch.id, cfg, lggr, ks, ch.Reader)
+	return &ch, nil
+}
+
+func (c *chain) ID() string {
+	return c.id
+}
+
+func (c *chain) Config() config.Config {
+	return c.cfg
+}
+
+func (c *chain) UpdateConfig(cfg db.ChainCfg) {
+	c.cfg.Update(cfg)
+}
+
+func (c *chain) TxManager() solana.TxManager {
+	return c.txm
+}
+
+func (c *chain) Reader() (solanaclient.Reader, error) {
+	return c.getClient()
+}
+
+// getClient returns a client, randomly selecting one from available and valid nodes
+func (c *chain) getClient() (solanaclient.ReaderWriter, error) {
+	var node db.Node
+	var client solanaclient.ReaderWriter
+	nodes, cnt, err := c.orm.NodesForChain(c.id, 0, math.MaxInt)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get nodes")
+	}
+	if cnt == 0 {
+		return nil, errors.New("no nodes available")
+	}
+	rand.Seed(time.Now().Unix()) // seed randomness otherwise it will return the same each time
+	// #nosec
+	index := rand.Perm(len(nodes)) // list of node indexes to try
+	for _, i := range index {
+		node = nodes[i]
+		// create client and check
+		client, err = c.verifiedClient(node)
+		// if error, try another node
+		if err != nil {
+			c.lggr.Warnw("failed to create node", "name", node.Name, "solana-url", node.SolanaURL, "error", err.Error())
+			continue
+		}
+		// if all checks passed, mark found and break loop
+		break
+	}
+	// if no valid node found, exit with error
+	if client == nil {
+		return nil, errors.New("no node valid nodes available")
+	}
+	c.lggr.Debugw("Created client", "name", node.Name, "solana-url", node.SolanaURL)
+	return client, nil
+}
+
+// verifiedClient returns a client for node or an error if the chain id does not match.
+func (c *chain) verifiedClient(node db.Node) (solanaclient.ReaderWriter, error) {
+	// create client
+	client, err := solanaclient.NewClient(node.SolanaURL, c.cfg, DefaultRequestTimeout, c.lggr.Named("Client-"+node.Name))
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create client")
+	}
+
+	// if endpoint has not been checked, fetch chainID
+	url := node.SolanaURL
+	c.chainIDLock.RLock()
+	id, exists := c.clientChainID[url]
+	c.chainIDLock.RUnlock()
+	if !exists {
+		id, err = client.ChainID()
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to fetch ChainID in checkClient")
+		}
+		c.chainIDLock.Lock()
+		c.clientChainID[url] = id
+		c.chainIDLock.Unlock()
+	}
+
+	// check chainID matches expected chainID
+	expectedID := strings.ToLower(c.id)
+	if id != expectedID {
+		return nil, fmt.Errorf("client returned mismatched chain id (expected: %s, got: %s): %s", expectedID, id, url)
+	}
+	return client, nil
+}
+
+func (c *chain) Start(ctx context.Context) error {
+	return c.StartOnce("Chain", func() error {
+		c.lggr.Debug("Starting")
+		c.lggr.Debug("Starting txm")
+		c.lggr.Debug("Starting balance monitor")
```

Comment:
These would be better at `trace` level

Reply:
Isn't debug level still the standard throughout the repo? Better to stay consistent here and change all at once, no?